### PR TITLE
refactor: apply new serialization

### PR DESCRIPTION
**Pull request by yangby-cryptape**
Size of `dao`: `24248 bytes -> 6992 bytes`.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -18,12 +18,16 @@ matrix:
   include:
     - name: Test
       script:
+        - cargo install moleculec
+        - make c/protocol_reader.h
         - make all-via-docker
-        - cargo test --all
+        - cargo test --tests
```
Comment (yangby-cryptape) [inline .travis.yml]: `cargo test --all` crashes randomly when linking doc-tests in my laptop, although this crate doesn't have any doc-tests.
(I spent a lot of time on this.) I think there is a bug in rust toolchain or any dependencies, so I updated this.

File: Cargo.toml [hunk 2]
```diff
@@ -20,12 +20,11 @@ faster-hex = "0.3"
 
 [dev-dependencies]
 byteorder = "1.3.1"
-ckb-core = { git = "https://github.com/nervosnetwork/ckb.git", rev = "9235afcc6cede6a097b747c3abbf9c4ceccf6583" }
-ckb-script = { git = "https://github.com/nervosnetwork/ckb.git", rev = "9235afcc6cede6a097b747c3abbf9c4ceccf6583" }
-ckb-crypto = { git = "https://github.com/nervosnetwork/ckb.git", rev = "9235afcc6cede6a097b747c3abbf9c4ceccf6583" }
-ckb-dao-utils = { git = "https://github.com/nervosnetwork/ckb.git", rev = "9235afcc6cede6a097b747c3abbf9c4ceccf6583" }
-ckb-hash = { git = "https://github.com/nervosnetwork/ckb.git", rev = "9235afcc6cede6a097b747c3abbf9c4ceccf6583" }
-numext-fixed-hash = "0.1.4"
+ckb-types = { git = "https://github.com/nervosnetwork/ckb.git", rev = "7e086b8e19542d85cd3dc5736ed17286695d31eb" }
```
Comment (yangby-cryptape) [inline Cargo.toml]: This commit will push into the `develop` branch.

File: c/ckb.mol [hunk 3]
```diff
@@ -0,0 +1,363 @@
+// Copy from ckb-types#656cb6e5
```
Comment (yangby-cryptape) [inline c/ckb.mol]: Only half of these structs are required.

When I finished the `import` statement grammar, we can split there structs into several schema files, then this crate could only include the schema file about blockchain.

File: deps/molecule/molecule.h [hunk 4]
```diff
@@ -0,0 +1,498 @@
+/* Copy from molecule v0.2.3 */
```
Comment (yangby-cryptape) [inline deps/molecule/molecule.h]: Some code can be deleted.

When I had time, I will split this file to two files: `molecule.h` and : `molecule-debug.h`, then this crate could only include the `molecule.h`.

File: c/secp256k1_blake160_sighash_all.c [hunk 5]
```diff
@@ -110,39 +91,50 @@ int main(int argc, char* argv[]) {
...
+      } else {
+        return ERROR_ENCODING;
+      }
+    }
+    witness_len = arg_res.attr;
+
+    bytes_res = mol_cut_bytes(&arg_res.pos);
+    if (bytes_res.code != 0) {
+      return ERROR_ENCODING;
+    } else if (bytes_res.pos.size < 65) {
```
Comment (jjyr) [inline c/secp256k1_blake160_sighash_all.c]: Define 65 as a constant `SIGNATURE_SIZE`

File: deps/molecule/molecule.h (hunk 4 above)
Comment (jjyr) [reply deps/molecule/molecule.h]: Should be a separate repo in someday?

File: c/secp256k1_blake160_sighash_all.c (hunk 5 above)
Comment (yangby-cryptape) [reply c/secp256k1_blake160_sighash_all.c]: I just inherited this from previous commits~ Maybe next PR?

File: deps/molecule/molecule.h (hunk 4 above)
Comment (doitian) [reply deps/molecule/molecule.h]: Molecule can add a sub command or option to dump the `.h` file.

File: c/secp256k1_ripemd160_sha256_sighash_all.c [hunk 6]
```diff
@@ -139,32 +119,37 @@ int main(int argc, char* argv[]) {
...
+      if (arg_res.attr != 2) {
+        return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+      } else {
+        return ERROR_ENCODING;
+      }
+    } else if (arg_res.attr != 2) {
       return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
     }
 
-    /* load pubkey and check pubkey hash */
```
Comment (driftluo) [inline c/secp256k1_ripemd160_sha256_sighash_all.c]: This comment should not be deleted

**Comment by yangby-cryptape** [comment]
> 1. The c methods return multiple values via a compound structure. I believe the pattern that returns the code and saves the result in the passed-in pointer is more popular.

I'm trying a new coding style ~ return value like `Result<(ptr,size),errno>` ~

> 2. `cut` looks like a two strong name for me. I prefer `slice`, `get_child` or `get_entity`.

Can be update, later.

> Name `slice_t` is better.

Maybe I should just use the same name as `bytes`~

**Pull request merged**